### ISSUE TITLE
Function 'motion_from_essential_choose_solution' is implemented incorrectly

### ISSUE THREAD

**Issue by Maltose0118** [kornia/kornia#2198]
### Describe the bug

In essential.py,  the two lines of code inside function 'motion_from_essential_choose_solution' are wrong:
`R_out = Rs[:, mask_indices][:, 0, 0]
t_out = ts[:, mask_indices][:, 0, 0]`

### Reproduction steps

```bash
Rs, ts and mask_indices are correct, but R_out and t_out are selected incorrectly.
```

### Expected behavior

The function selects wrong R, t decomposed by essential mat.

### Environment

```shell
python-3.8.13
pytorch-1.8.2
kornia-0.6.8
```

**Comment by edgarriba** [comment]
@Maltose0118 thanks to report. Can you send a fix with some tests showing where it fails ?

**Comment by edgarriba** [comment]
If anyone is interested in this, feel free to re-open the ticket. Thanks!

**Comment by ducha-aiki** [comment]
The 2023 report is reproducible on `main` today: for a batch, `motion_from_essential_choose_solution` applies element 0's candidate index to every element.

`mask_indices` has shape `(B, 1)` (`essential.py:756`), so `Rs[:, mask_indices]` indexes dim 1 of every element with every element's index, giving `(B, B, 1, 3, 3)`, and `[:, 0, 0]` keeps the index chosen for element 0 (`essential.py:759-761`, the lines quoted above; `t` and `points3d` alike). The repro needs a batch whose elements have different true-candidate indices: `E` and `-E` describe the same motion, but `motion_from_essential` lists the true pose at index 2 for `E` and at index 0 for `-E`. Each alone gives the right pose; as the batch `[E, -E]` with the same correspondences, element 1 gets a rotation 2.0 (max-abs entry difference) from the truth and points at negative depth. A batch whose elements share the index, such as `[E, E]`, looks correct. `TestMotionFromEssentialChooseSolution` checks values only for a batch of one (`test_two_view`); its batched `test_shape` checks shapes.

## Repro

```python
import torch
from kornia.geometry.conversions import axis_angle_to_rotation_matrix
from kornia.geometry.epipolar import essential_from_Rt, motion_from_essential, motion_from_essential_choose_solution

K1 = torch.tensor([[[800.0, 0, 320], [0, 760, 200], [0, 0, 1]]])
K2 = torch.tensor([[[700.0, 0, 300], [0, 740, 240], [0, 0, 1]]])
R = axis_angle_to_rotation_matrix(torch.tensor([[0.1, -0.2, 0.05]]))
t = torch.tensor([[[0.5], [0.1], [0.02]]])
torch.manual_seed(0)
X = torch.rand(12, 3) * 2 - 1
X[:, 2] += 5
x1 = (K1[0] @ X.T).T
x2 = (K2[0] @ (R[0] @ X.T + t[0])).T
x1, x2 = (x1[:, :2] / x1[:, 2:])[None], (x2[:, :2] / x2[:, 2:])[None]
E = essential_from_Rt(torch.eye(3)[None], torch.zeros(1, 3, 1), R, t)

# E and -E describe the same motion; motion_from_essential lists the true pose at a different index for each
for name, e in [("E", E), ("-E", -E)]:
    Rs, ts = motion_from_essential(e)
    idx = [i for i in range(4) if (Rs[0, i] - R[0]).abs().max() < 1e-5 and (ts[0, i] - t[0] / t.norm()).abs().max() < 1e-5]
    Ri, ti, Xi = motion_from_essential_choose_solution(e, K1, K2, x1, x2)
    print(f"{name:2s} alone:    true candidate index {idx}, |R - R_true| = {(Ri - R).abs().max().item():.2g}, "
          f"t = {[round(v, 4) for v in ti.flatten().tolist()]}, min depth = {Xi[..., 2].min().item():.3g}")

# the same two problems as one batch
Rb, tb, Xb = motion_from_essential_choose_solution(torch.cat([E, -E]), K1.expand(2, 3, 3), K2.expand(2, 3, 3),
                                                   x1.expand(2, -1, -1), x2.expand(2, -1, -1))
for i, name in enumerate(["E", "-E"]):
    print(f"batch [E, -E] element {i} ({name:2s}): |R - R_true| = {(Rb[i] - R[0]).abs().max().item():.2g}, "
          f"t = {[round(v, 4) for v in tb[i].flatten().tolist()]}, min depth = {Xb[i, :, 2].min().item():.3g}")
```

Output:

```text
E  alone:    true candidate index [2], |R - R_true| = 6e-08, t = [0.9798, 0.196, 0.0392], min depth = 7.96
-E alone:    true candidate index [0], |R - R_true| = 6e-08, t = [0.9798, 0.196, 0.0392], min depth = 7.96
batch [E, -E] element 0 (E ): |R - R_true| = 6e-08, t = [0.9798, 0.196, 0.0392], min depth = 7.96
batch [E, -E] element 1 (-E): |R - R_true| = 2, t = [0.9798, 0.196, 0.0392], min depth = -80.9
```

kornia `main` 7ddf731bd, torch 2.14.0 (and 2.5.1: same), CPU float32, macOS arm64.

## Classification

Silent wrong result. No idiom of 1–9; candidate: advanced indexing with a `(B, 1)` index on dim 1 broadcasts across the batch.

## Disposition

Focused fix, welcome as a PR: gather per element, e.g. `b = torch.arange(B, device=...)`, `R_out = Rs[b, mask_indices[:, 0]]` (and `t_out`, `points3d_out`), with a test on the batch `[E, -E]`. Unbatched calls and batches of one are unchanged.

Found by the batch-8 conventions audit (two-view geometry); the `test_wart_motion_from_essential_choose_solution_batch_index_2198` pin in the docs PR flips when the fix lands.

Related: #4879 (same function, no point passes the depth test).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
